### PR TITLE
Nuked TGenotype

### PR DESCRIPTION
**Pull request by cseed**
TGenotype gone and the Genotype class is gone.  The few remaining legacy genotypes are stored generically as annotations.  The changes are primarily what's naturally forced.

Main unforced change was to revise the way VSM.gen apportions sizes.  Time seems roughly the same and the new generator turned up a few bugs so I think it is better.

@tpoterba Sorry, this is a monster.  I can try to break it up if needed.

### CODE REVIEW

File: python/hail/dataset.py [hunk 1]
```diff
@@ -1664,11 +1643,9 @@ def export_vcf(self, output, append_to_header=None, parallel=False):
         the elements of Set[String] ``va.filters`` as FILTERS,
         and the value of Float64 ``va.qual`` as QUAL. No other variant annotations are exported.
         
-        The FORMAT field is generated from the genotype type, which must 
-        be :py:class:`~hail.expr.TGenotype` or :py:class:`~hail.expr.TStruct`. If the type is
-        :py:class:`~hail.expr.TGenotype`, then the FORMAT fields will be GT, AD, DP, GQ, and PL.
-        If the type is :py:class:`~hail.expr.TStruct`,
-        then the exported FORMAT fields will be the names of each field of the Struct.
+        The FORMAT field is generated from the genotype type, which
```
Comment (tpoterba) [inline python/hail/dataset.py]: funky language here, something got deleted and not clarified.

File: python/hail/dataset.py [hunk 2]
```diff
@@ -2152,7 +2130,7 @@ def filter_samples_expr(self, expr, keep=True):
         - ``s`` (*Sample*): sample
         - ``sa``: sample annotations
         - ``global``: global annotations
-        - ``gs`` (*Aggregable[Genotype]*): aggregable of :ref:`genotype` for sample ``s``
+        - ``gs`` (*Aggregable[Genotype]*): aggregable of genotype for sample ``s``
```
Comment (tpoterba) [inline python/hail/dataset.py]: aggregable[genotype] is wrong

Comment (tpoterba) [reply python/hail/dataset.py]: this is found a bunch of places, actually.

Comment (tpoterba) [reply python/hail/dataset.py]: Maybe 
```
- ``gs`` (*Aggregable[Struct]*): aggregable of genotype structs for sample ``s``
```

File: python/hail/docs/overview.rst [hunk 3]
```diff
@@ -19,7 +19,7 @@ Variant Dataset (VDS)
 
 Hail represents a genetic data set as a matrix where the rows are keyed by
 :ref:`variant(gr)` objects, the columns are keyed by samples, and each cell is a
-:ref:`genotype` object. :ref:`variant(gr)` objects and :ref:`genotype` objects each
+genotype object. :ref:`variant(gr)` objects and genotype objects each
```
Comment (tpoterba) [inline python/hail/docs/overview.rst]: genotype object is now an outdated term

File: python/hail/representation/genotype.py [hunk 4]
```diff
@@ -346,6 +34,12 @@ def __str__(self):
     def __repr__(self):
         return 'Call(gt=%s)' % self._gt
 
+    def __eq__(self, other):
+        return self.gt == other.gt
```
Comment (tpoterba) [inline python/hail/representation/genotype.py]: `return isinstance(other, Call) and self.gt == other.gt`

File: python/hail/representation/genotype.py [hunk 5]
```diff
@@ -346,6 +34,12 @@ def __str__(self):
...
+
+    def __hash__(self):
+        return hash(self.gt)
```
Comment (tpoterba) [inline python/hail/representation/genotype.py]: can we add some magic number here so hash(Call(1)) isn't hash(1)?

File: python/hail/tests/tests.py [hunk 6]
```diff
@@ -329,7 +316,8 @@ def test_dataset(self):
 
         sample2.split_multi().impute_sex().variant_schema
 
-        self.assertTrue(isinstance(sample2.genotype_schema, TGenotype))
+        # FIXME htsGenotypeSchema
```
Comment (tpoterba) [inline python/hail/tests/tests.py]: can you explain this a bit?

File: python/hail2/tests.py [hunk 7]
```diff
@@ -170,17 +170,17 @@ def test_select(self):
         self.assertEqual(kt.select(False, kt.a, *kt.g, foo=kt.a + kt.b - kt.c - kt.d).columns, ['a', 'x', 'y', 'foo'])
 
     def test_aggregate(self):
-        schema = TStruct(['status', 'gt', 'qPheno'],
-                         [TInt32(), TGenotype(), TInt32()])
+        schema = TStruct(['status', 'GT', 'qPheno'],
```
Comment (tpoterba) [inline python/hail2/tests.py]: the py-expr-2 rebase is going to be very real.

File: src/main/scala/is/hail/methods/ImputeSexPlink.scala [hunk 8]
```diff
@@ -17,19 +17,27 @@ object ImputeSexPlink {
       case TVariant(x, _) => x.asInstanceOf[GenomeReference]
     }
 
+    vsm.typecheck()
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/ImputeSexPlink.scala]: seems like you meant to remove these?

File: src/main/scala/is/hail/stats/CallStatsCombiner.scala [hunk 9]
```diff
@@ -15,18 +15,19 @@ object CallStats {
 
 case class CallStats(alleleCount: IndexedSeq[Int], alleleFrequency: Option[IndexedSeq[Double]], alleleNumber: Int,
   genotypeCount: IndexedSeq[Int]) {
+  println(alleleNumber, genotypeCount.toSeq, genotypeCount.sum)
```
Comment (tpoterba) [inline src/main/scala/is/hail/stats/CallStatsCombiner.scala]: remove

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 10]
```diff
@@ -674,20 +345,25 @@ object Genotype {
...
         }
       }
-      val g = Genotype(gt, ad, dp.map(_ + ad.map(_.sum).getOrElse(0)), gq, pl)
-      g.check(nAlleles)
+      val g = Annotation(gt.orNull,
+        ad.map(a => a: IndexedSeq[Int]).orNull,
+        dp.map(_ + ad.map(_.sum).getOrElse(0)).orNull,
+        gq.orNull,
+        pl.map(a => a: IndexedSeq[Int]).orNull)
+      // FIXME
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/Genotype.scala]: ?

File: src/main/scala/is/hail/variant/HTSGenotypeView.scala [hunk 11]
```diff
@@ -8,17 +8,13 @@ import is.hail.utils._
...
-      case t => fatal(s"invalid genotype representation: $t, expect TGenotype or TStruct")
-    }
+    new StructGenotypeView(rowSignature)
   }
 
   val tArrayInt32 = TArray(!TInt32())
 }
 
-// FIXME: This is removed, and StructGenotypeView becomes the only genotype view when TGenotype is removed
+// FIXME remove
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/HTSGenotypeView.scala]: do you want to do this cleanup in a followup PR?

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 12]
```diff
@@ -290,51 +282,11 @@ class VariantDatasetFunctions(private val vsm: VariantSampleMatrix) extends AnyV
 
   def filterAlleles(filterExpr: String, variantExpr: String = "",
     keep: Boolean = true, subset: Boolean = true, leftAligned: Boolean = false, keepStar: Boolean = false): VariantSampleMatrix = {
+    if (!vsm.genotypeSignature.isOfType(Genotype.htsGenotypeType))
+      fatal(s"filter_alleles: genotype_schema must be TGenotype or the HTS genotype schema, found: ${ vsm.genotypeSignature }")
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: tgenotype reference

File: src/test/resources/filter.vcf [hunk 13]
```diff
@@ -1,11 +1,9 @@
 ##fileformat=VCFv4.1
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
```
Comment (tpoterba) [inline src/test/resources/filter.vcf]: changes to this file are to make it compatible with the HTS schema?

Comment (tpoterba) [reply src/test/resources/filter.vcf]: to this and all the following files, rather.

Comment (cseed) [reply src/test/resources/filter.vcf]: Yes, exactly.

File: python/hail/dataset.py (hunk 2 above)
Comment (cseed) [reply python/hail/dataset.py]: It doesn't have to be a struct.  I'm trying to sloppily use "genotype" to mean whatever the cell in the matrix is so I don't have to revise all the documentation when it is going to get revised again for the new python API.  Another idea: I could still write `Aggregable[genotype]`, but have the genotype link out to `genotype_schema`.

Comment (tpoterba) [reply python/hail/dataset.py]: Alright, yeah. Don't worry about this.

File: python/hail/tests/tests.py (hunk 6 above)
Comment (cseed) [reply python/hail/tests/tests.py]: Yeah, I wanted to check the genotype_schema is hts_genotype_schema, but I need to export it to python.  Will do.

**Comment by cseed** [comment]
I addressed some comments.  I still need to:
 - expose hts_genotype_schema in python, and
 - figure out what to rewrite instead of "genotype" in the VariantDataset docs.

**Comment by cseed** [comment]
I pushed more changes to remove is{Not}Called and renamed isCalledNonRef => isNonRef.

I will address comments and rebase after the rest of the stack goes in.

**Comment by cseed** [comment]
Added `Type.hts_schema` in Python.  I'm not sure that's the right place for it and I couldn't make it a property because classes have properties (although it might be possible to do it with a property on the metaclass).

**Comment by tpoterba** [comment]
OK, will take another look.

File: src/main/scala/is/hail/expr/Relational.scala [hunk 14]
```diff
@@ -185,11 +185,11 @@ object MatrixValue {
...
             rdd.orderedPartitioner.rangeBounds.map(b => Row(b)))),
         rdd.mapPartitions { it =>
-          val region = MemoryBuffer()
-          val rvb = new RegionValueBuilder(region)
-          val rv = RegionValue(region)
-
           it.map { case (v, (va, gs)) =>
+            val region = MemoryBuffer()
+            val rvb = new RegionValueBuilder(region)
+            val rv = RegionValue(region)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Relational.scala]: why this change? Wasn't it working with mutation?

File: src/main/scala/is/hail/variant/Genotype.scala [hunk 15]
```diff
@@ -54,71 +49,6 @@ class GTPair(val p: Int) extends AnyVal {
...
-    b.append(Genotype.gq(this).map(_.toString).getOrElse("."))
-    b += ':'
-    b.append("PL=" + Genotype.pl(this).map(_.mkString(",")).getOrElse("."))
-
-    b.result()
-  }
-}
-
 object Genotype {
   val htsGenotypeType: TStruct = TStruct(
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/Genotype.scala]: This doesn't seem like the long-term design we want, because in order to run variantQC, you don't need something that is exactly GT/AD/DP/GQ/PL, but just something that contains those fields in any order (and could contain others). So we wouldn't want to check equality, but instead to check membership. Is this a temporary workaround to keep stuff working with the tests?

Comment (cseed) [reply src/main/scala/is/hail/variant/Genotype.scala]: This isn't used in variantQC.  It used in split_multi and related routines.

split_multi requires exactly this schema (although in an arbitrary order is fine), because it has a fixed schema on output and will delete any other genotype fields present.  Although the input can be in any order, the output is in the standard order.  I have to do this because we can't put an insertion inside a let and I need, for example, to compute GQ  from PL (without recording PL twice).  Also, instantiating the entire struct is way faster than doing 5 insertions (although that's a short term concern).

So ... I could make split_multi require the right fields but in any order.

File: src/main/scala/is/hail/expr/Relational.scala (hunk 14 above)
Comment (cseed) [reply src/main/scala/is/hail/expr/Relational.scala]: Didn't mean to put this in, I'll remove.

File: src/main/scala/is/hail/variant/Genotype.scala (hunk 15 above)
Comment (tpoterba) [reply src/main/scala/is/hail/variant/Genotype.scala]: Alright. Let's revisit this again in a month or so, to evaluate whether it's still the right thing. Seems fine for now.

File: src/main/scala/is/hail/expr/Relational.scala (hunk 14 above)
Comment (cseed) [reply src/main/scala/is/hail/expr/Relational.scala]: Fixed.

**Comment by cseed** [comment]
Rebased.  Despite some linger issues (docs, handling the HTS schema) I think this is ready for another look.

**Comment by tpoterba** [comment]
I'm not too concerned about docs, since it's all going to be revisited soon.

File: python/hail/tests/tests.py [hunk 16]
```diff
@@ -33,6 +33,14 @@ def float_eq(x, y, tol=10 ** -6):
     return abs(x - y) < tol
 
 class ContextTests(unittest.TestCase):
+    # FIXME remove when attributes get nuked
+    def assert_equal_struct(self, s1, s2):
```
Comment (tpoterba) [inline python/hail/tests/tests.py]: wait, hold off. Attributes have been nuked.

Comment (cseed) [reply python/hail/tests/tests.py]: Fixed!

**Comment by cseed** [comment]
Python tests fixed.
